command/agent_test.go

```diff
@@ -1234,3 +1228,462 @@ func makeTempFile(t *testing.T, name, contents string) string {
+				// likely fail the test
+				tick := time.Tick(1 * time.Second)
+				timeout := time.After(10 * time.Second)
+				var err error
+				for {
+					select {
+					case <-timeout:
+						return fmt.Errorf("timed out waiting for templates to render, last error: %v", err)
+					case <-tick:
+					}
+					// Check for files rendered in the directory and break
+					// early for shutdown if we do have all the files
+					// rendered
+
+					//----------------------------------------------------
+					// Perform the tests
+					//----------------------------------------------------
+
+					if numFiles := testListFiles(t, tmpDir, ".json"); numFiles != 1 {
+						err = fmt.Errorf("expected 1 template, got (%d)", numFiles)
+						continue
+					}
+
+					fileName := filepath.Join(tmpDir, "render_0.json")
+					var c []byte
+					c, err = ioutil.ReadFile(fileName)
+					if err != nil {
+						continue
+					}
+					if string(c) != templateRendered(0) {
+						err = fmt.Errorf("expected='%s', got='%s'", templateRendered(0), string(c))
+						continue
+					}
+					return nil
+				}
+			}
+
+			err = verify()
+			close(cmd.ShutdownCh)
+			wg.Wait()
+
+			switch {
+			case (code != 0 || err != nil) && tc.expectError:
+			case code == 0 && err == nil && !tc.expectError:
+			default:
+				t.Fatalf("%s expectError=%v error=%v code=%d", tcname, tc.expectError, err, code)
+			}
+
+		})
+	}
+}
+
+// prepAgentApproleKV configures a Vault instance for approle authentication,
+// such that the resulting token will have global permissions across /kv
+// and /secret mounts.  Returns the auto_auth config stanza to setup an Agent
+// to connect using approle.
+func prepAgentApproleKV(t *testing.T, client *api.Client) (string, func()) {
+	t.Helper()
+
+	policyAutoAuthAppRole := `
+path "/kv/*" {
+	capabilities = ["create", "read", "update", "delete", "list"]
+}
+path "/secret/*" {
+	capabilities = ["create", "read", "update", "delete", "list"]
+}
+`
+	// Add an kv-admin policy
+	if err := client.Sys().PutPolicy("test-autoauth", policyAutoAuthAppRole); err != nil {
+		t.Fatal(err)
+	}
+
+	// Enable approle
+	err := client.Sys().EnableAuthWithOptions("approle", &api.EnableAuthOptions{
+		Type: "approle",
+	})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	_, err = client.Logical().Write("auth/approle/role/test1", map[string]interface{}{
+		"bind_secret_id": "true",
+		"token_ttl":      "3s",
+		"token_max_ttl":  "10s",
+		"policies":       []string{"test-autoauth"},
+	})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	resp, err := client.Logical().Write("auth/approle/role/test1/secret-id", nil)
+	if err != nil {
+		t.Fatal(err)
+	}
+	secretID := resp.Data["secret_id"].(string)
+	secretIDFile := makeTempFile(t, "secret_id.txt", secretID+"\n")
+
+	resp, err = client.Logical().Read("auth/approle/role/test1/role-id")
+	if err != nil {
+		t.Fatal(err)
+	}
+	roleID := resp.Data["role_id"].(string)
+	roleIDFile := makeTempFile(t, "role_id.txt", roleID+"\n")
+
+	return fmt.Sprintf(`
```

Comment:
Nit: this is a bit hard to read, wondering if you can assign vars and return them here instead?

Reply:
Done